tests/test_branch_stats.py

```diff
@@ -287,54 +460,99 @@ def f(x):
-        samples = random.sample(ts.samples(), 3)
-        A = [[samples[0]], samples[1:3]]
-
-        def f(x):
-            return float(((x[0] == 1) and (x[1] == 0)) or ((x[0] == 0) and (x[1] == 2)))
+        samples = random.sample(ts.samples(), 6)
+        A = [[samples[0]], [samples[1]], [samples[2], samples[3]],
+             [samples[4], samples[5]]]
+        windows = [0.0, ts.sequence_length/20, ts.sequence_length/2, ts.sequence_length]
+        here_values = [branch_length_Y(ts, A[0], A[1], A[2],
+                                       begin=windows[k], end=windows[k+1])
+                       for k in range(len(windows)-1)]
+        here_values_2 = [branch_length_Y(ts, A[1], A[2], A[3],
+                                         begin=windows[k], end=windows[k+1])
+                         for k in range(len(windows)-1)]
+        ts_values = ts.Y(A[0:3], windows)
+        ts_vector_values = ts.Y_vector(A, windows, [[0, 1, 2], [1, 2, 3], [0, 2, 1]])
 
-        self.assertAlmostEqual(
-                ts.branch_stats(A, f),
-                branch_length_Y(ts, A[0][0], A[1][0], A[1][1]))
-        self.assertAlmostEqual(
-                branch_stats_node_iter(ts, A, f, method='length'),
-                branch_length_Y(ts, A[0][0], A[1][0], A[1][1]))
+        self.assertListAlmostEqual(
+                [x[0] for x in ts_values],
+                here_values)
+        self.assertListAlmostEqual(
+                [x[1] for x in ts_vector_values],
+                here_values_2)
+        self.assertListAlmostEqual(
+                [x[2] for x in ts_vector_values],
+                here_values)
 
     def check_f4_stat(self, ts):
-        samples = random.sample(ts.samples(), 4)
-        A_zero = [[samples[0]], [samples[0]], [samples[1]], [samples[1]]]
-        A_f1 = [[samples[0]], [samples[1]], [samples[0]], [samples[1]]]
-        A_one = [[samples[0]], [samples[1]], [samples[2]], [samples[3]]]
+        A_zero = [[x] for x in random.sample(ts.samples(), 4)]
         A_many = [random.sample(ts.samples(), 3),
                   random.sample(ts.samples(), 3),
                   random.sample(ts.samples(), 3),
                   random.sample(ts.samples(), 3)]
-        for A in (A_zero, A_f1, A_one, A_many):
+        A_list = A_zero + A_many
+        windows = [0.0, ts.sequence_length/2.0, ts.sequence_length]
+        indices = [(0, 1, 2, 3), (0, 1, 4, 5), (4, 5, 6, 7)]
+        ts_vector = ts.f4_vector(A_list, windows, indices)
+        for k in range(len(indices)):
+            index_list = indices[k]
+            A = [A_list[i] for i in index_list]
 
             def f(x):
                 return ((float(x[0])/len(A[0])-float(x[1])/len(A[1]))
                         * (float(x[2])/len(A[2])-float(x[3])/len(A[3])))
 
-            self.assertAlmostEqual(
-                    ts.branch_stats(A, f),
-                    branch_length_f4(ts, A[0], A[1], A[2], A[3]))
+            here_values = [branch_length_f4(ts, A[0], A[1], A[2], A[3],
```

Comment:
Ah good catch. And nice to know about the variable scoping difference w/r/t list comprehensions